### ISSUE TITLE
Using :proxy can leave zombie processes behind in case of timeout

### ISSUE THREAD

**Issue by DimitriosLisenko** [net-ssh/net-ssh#557]
### Preliminary steps
Try to open a connection through a proxy.

### Expected behaviour
Reap proxy command child process if timeout occurs.

### Actual behaviour
If either the proxy command connection times out, a zombie process is left behind.

### System configuration
net-ssh version: 4.2.0
Ruby version: 2.3.1p112

### Example App
Previously there was an issue #526 which addresses a zombie process left behind in case of an error in the proxy command. This is still an issue in case of a timeout or exception when the proxy command happens - to reproduce I used the following script:

```
gem 'net-ssh', '= 4.2.0'
require 'net/ssh'
require 'net/ssh/proxy/command'
puts Net::SSH::Version::CURRENT

begin
  proxy = Net::SSH::Proxy::Command.new('ssh -W %h:%p localhost')
  Net::SSH.start('localhost', ENV['USER'], :verbose => 0, :proxy => proxy) do |ssh|
    puts ssh.exec! 'hostname'
  end
rescue => e
  puts `ps -ef | grep #{Process.pid}`
end
```

I then ran `sudo iptables -A INPUT -p tcp --dport 22 -j DROP` on the machine to block all traffic to port 22 and then ran `load "./reproducer.rb"` from the `irb` console.

Watching the result using `watch -n 0.1 'ps -elf --forest'`, you can see that after the `Proxy::Command` timeout (60 seconds), the proxy command gets orphaned, and then if you keep watching, it will eventually become a zombie.

```
F S UID        PID  PPID  C PRI  NI ADDR SZ WCHAN  STIME TTY          TIME CMD
4 S root      1562     0  0  80   0 -  4560 -      11:32 pts/2    00:00:00 /bin/bash
0 S root     11915  1562  0  80   0 - 20120 -      11:39 pts/2    00:00:00  \_ irb
4 S root     15263 11915  0  80   0 - 11197 -      11:42 pts/2    00:00:00      \_ ssh -W localhost:22 localhost

# Some time later
F S UID        PID  PPID  C PRI  NI ADDR SZ WCHAN  STIME TTY          TIME CMD
4 S root      1562     0  0  80   0 -  4560 -      11:32 pts/2    00:00:00 /bin/bash
0 S root     11915  1562  0  80   0 - 20120 -      11:39 pts/2    00:00:00  \_ irb
4 Z root     15263 11915  0  80   0 -     0 -      11:42 pts/2    00:00:00      \_ [ssh] <defunct>
```

I have created a possible fix at #556 - it involves handling the child process in the `Proxy::Command` class, since in the timeout/exception case the IO is not returned, and so can't be handled by the `Transport::Session` class. Another option is to maybe set `autoclose` to `true` on the pipe.

Open to discussion - perhaps instead of just closing the IO stream in this scenario (which would hang) the pipe should instead be killed first via `Process.kill("INT", io.pid)` and then `io.close` should be called to avoid the process becoming defunct.

**Comment by mfazekas** [comment]
Thanks much for the detailed report and esp for the script to reproduce.

👍

**Comment by WyseNynja** [comment]
https://github.com/net-ssh/net-ssh/pull/556 isn't a clean merge. Can you rebase, please? This zombie leak is tolerable on our developer laptops, but we have a service that uses vagrant a lot more and these zombies start to become an issue there.

**Comment by mfazekas** [comment]
Should be closed by #580
